setup.py

```diff
@@ -106,20 +106,35 @@ def add_optimization_flags(self):
         """
 
         env = os.environ.copy()
-        if not setupext.config.getboolean('libs', 'enable_lto', fallback=True):
-            return env
         if sys.platform == 'win32':
             return env
-
-        cppflags = []
-        if 'CPPFLAGS' in os.environ:
-            cppflags.append(os.environ['CPPFLAGS'])
-        cxxflags = []
-        if 'CXXFLAGS' in os.environ:
-            cxxflags.append(os.environ['CXXFLAGS'])
-        ldflags = []
-        if 'LDFLAGS' in os.environ:
-            ldflags.append(os.environ['LDFLAGS'])
+        enable_lto = setupext.config.getboolean('libs', 'enable_lto',
+                                                fallback=None)
+
+        def prepare_flags(name, enable_lto):
+            """
+            Prepare *FLAGS from the environment.
+
+            If set, return them, and also check whether LTO is disabled in each
+            one, raising an error if Matplotlib config explicitly enabled LTO.
+            """
+            flags = []
+            if name in os.environ:
+                flags.append(os.environ[name])
+                if '-fno-lto' in os.environ[name]:
+                    if enable_lto is True:
+                        raise ValueError('Configuration enable_lto=True, but '
+                                         '{0} contains -fno-lto'.format(name))
+                    enable_lto = False
+            return flags, enable_lto
```

Comment:
perhaps something like
```python
def prepare_flags(name):
    nonlocal enable_lto
    if name in os.environ:
        if "-fno-lto" in os.environ[name]:
            if enable_lto is True: raise ValueError(...)
            enable_lto = False
        return [os.environ[name]]
    else:
        return []
```
?

Reply:
I can inline the `return`, but `nonlocal` breaks Python 2 parsing for this file, so can't do that.

Reply:
fair